### PR TITLE
test(annotation): cover tasks status validation functions

### PR DESCRIPTION
**Pull request by BogicevicMilan**

### CODE REVIEW

File: annotation/tests/test_tasks_services.py [hunk 1]
```diff
@@ -794,9 +790,9 @@ def test_filter_tasks_db_no_files_or_jobs(
 )
 def test_create_tasks(
     mock_session: Mock,
-    tasks: List[Dict[str, Union[Set[int], int]]],
+    tasks: Dict[str, Union[int, set]],
     job_id: int,
-    expected_user_ids: Set[int],
+    expected_user_ids: set,
```
Comment (cakeinsauce) [inline annotation/tests/test_tasks_services.py:795]: explicit data type for containers

File: annotation/tests/test_tasks_services.py [hunk 2]
```diff
@@ -890,3 +886,146 @@ def test_get_task_revisions(
...
+    mock_session.query.return_value.filter.assert_called_once()
+    mock_unblocked_tasks.update.assert_called_once_with(
+        {"status": TaskStatusEnumSchema.ready},
+        synchronize_session=False,
+    )
+    assert result == [mock_task]
+
+
+def test_get_task_stats_by_id(mock_session: Mock):
+    mock_stats = MagicMock(spec=AnnotationStatistics)
```
Comment (cakeinsauce) [inline annotation/tests/test_tasks_services.py:925]: use real data type

File: annotation/tests/test_tasks_services.py [hunk 3]
```diff
@@ -890,3 +886,146 @@ def test_get_task_revisions(
...
+    result = services.get_task_stats_by_id(mock_session, task_id=1)
+    mock_session.query.assert_called_once_with(AnnotationStatistics)
+    mock_query.filter.assert_called_once()
+    mock_query.filter.return_value.first.assert_called_once()
+    assert result == mock_stats
+
+
+def test_add_task_stats_record_existing_stats(mock_session: Mock):
+    task_id = 1
+    mock_stats_input = MagicMock(spec=AnnotationStatisticsInputSchema)
```
Comment (cakeinsauce) [inline annotation/tests/test_tasks_services.py:938]: use real data type

File: annotation/tests/test_tasks_services.py [hunk 4]
```diff
@@ -890,3 +886,146 @@ def test_get_task_revisions(
...
+    mock_stats_input.event_type = "open"
+    mock_stats_db = MagicMock(spec=AnnotationStatistics)
```
Comment (cakeinsauce) [inline annotation/tests/test_tasks_services.py:940]: use real data type

File: annotation/tests/test_tasks_services.py [hunk 5]
```diff
@@ -890,3 +886,146 @@ def test_get_task_revisions(
...
+        mock_stats_db.field2 = "value2"
+        mock_stats_db.updated = datetime.utcnow()
+        mock_session.add.assert_called_once_with(mock_stats_db)
+        mock_session.commit.assert_called_once()
+        assert result == mock_stats_db
+
+
+def test_add_task_stats_record(mock_session: Mock):
+    task_id = 1
+    mock_stats_input = MagicMock(spec=AnnotationStatisticsInputSchema)
```
Comment (cakeinsauce) [inline annotation/tests/test_tasks_services.py:961]: use real data type

File: annotation/tests/test_tasks_services.py [hunk 6]
```diff
@@ -890,3 +886,146 @@ def test_get_task_revisions(
...
+        mock_get_task_stats_by_id.assert_called_once_with(
+            mock_session, task_id
+        )
+        mock_session.add.assert_not_called()
+        mock_session.commit.assert_not_called()
+
+
+def test_add_task_stats_record_setattr(mock_session: Mock):
+    task_id = 1
+    mock_stats_input = MagicMock(spec=AnnotationStatisticsInputSchema)
```
Comment (cakeinsauce) [inline annotation/tests/test_tasks_services.py:979]: use real data type

File: annotation/tests/test_tasks_services.py [hunk 7]
```diff
@@ -890,3 +886,146 @@ def test_get_task_revisions(
...
+    mock_stats_input.dict.return_value = {
+        "field1": "new_value1",
+        "field2": "new_value2",
+    }
+    mock_stats_db = MagicMock(spec=AnnotationStatistics)
```
Comment (cakeinsauce) [inline annotation/tests/test_tasks_services.py:984]: use real data type

File: annotation/tests/test_tasks_services.py [hunk 8]
```diff
@@ -890,3 +886,146 @@ def test_get_task_revisions(
...
+        assert mock_stats_db.field2 == "new_value2"
+        assert mock_stats_db.updated is not None
+        mock_session.add.assert_called_once_with(mock_stats_db)
+        mock_session.commit.assert_called_once()
+        assert result == mock_stats_db
+
+
+def test_add_task_stats_record_create_new(mock_session: Mock):
+    task_id = 1
+    mock_stats_input = MagicMock(spec=AnnotationStatisticsInputSchema)
```
Comment (cakeinsauce) [inline annotation/tests/test_tasks_services.py:1007]: use real data type

File: annotation/tests/test_tasks_services.py [hunk 9]
```diff
@@ -890,3 +886,146 @@ def test_get_task_revisions(
...
+    mock_stats_input.event_type = "open"
+    mock_stats_input.dict.return_value = {
+        "field1": "value1",
+        "field2": "value2",
+    }
+    mock_stats_db = MagicMock()
```
Comment (cakeinsauce) [inline annotation/tests/test_tasks_services.py:1013]: use real data type

File: annotation/tests/test_tasks_services.py (hunk 1 above)
Comment (BogicevicMilan) [reply annotation/tests/test_tasks_services.py:795]: It was okay, before merging. I'll change it again.

Comment (cakeinsauce) [reply annotation/tests/test_tasks_services.py]: still there

File: annotation/tests/test_tasks_services.py [hunk 10]
```diff
@@ -890,3 +887,144 @@ def test_get_task_revisions(
...
+    ) as mock_get_task_stats_by_id:
+        result = services.add_task_stats_record(
+            mock_session, task_id, mock_stats_input
+        )
+        mock_get_task_stats_by_id.assert_called_once_with(
+            mock_session, task_id
+        )
+        mock_stats_db.field1 = "value1"
+        mock_stats_db.field2 = "value2"
+        mock_stats_db.updated = datetime.utcnow()
```
Comment (cakeinsauce) [inline annotation/tests/test_tasks_services.py:955]: provide these fields in the constructor of the object

File: annotation/tests/test_tasks_services.py [hunk 11]
```diff
@@ -890,3 +887,144 @@ def test_get_task_revisions(
...
+        mock_session.add.assert_called_once_with(mock_stats_db)
+        mock_session.commit.assert_called_once()
+        assert result == mock_stats_db
+
+
+def test_add_task_stats_record(mock_session: Mock):
+    task_id = 1
+    mock_stats_input = AnnotationStatisticsInputSchema(
+        event_type=AnnotationStatisticsEventEnumSchema.closed
+    )
```
Comment (cakeinsauce) [inline annotation/tests/test_tasks_services.py:965]: you dont use this, remove if not needed

Comment (BogicevicMilan) [reply annotation/tests/test_tasks_services.py:965]: It is there because I copied it from prev example.

**Review by cakeinsauce — APPROVED** [review]
lgtm!

**Pull request merged**